### PR TITLE
enable data preload write through

### PR DESCRIPTION
**Pull request by Clark0**
### Changelog category :
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Write through cache strategy, 
add new settings to enable writing to disk cache while inserting data
* enable_preload_parts
* enable_async_preload_parts
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

### CODE REVIEW

File: src/Storages/MergeTree/MergeTreeSettings.h [hunk 1]
```diff
@@ -222,6 +222,7 @@ struct Settings;
     M(Bool, enable_persistent_checksum, false, "Persist checksums of part in memory. If set to false, checksums will be managed by a global cache to save memory.", 0) \
     \
     M(Bool, enable_local_disk_cache, true, "Enable local disk cache", 0) \
+    M(Bool, enable_preload_parts, false, "Enable preload parts", 0) \
```
[the remark this answers is not in the document]
Comment (hustnn) [reply src/Storages/MergeTree/MergeTreeSettings.h:225]: oh, isee, it reuse existing session resource logic

File: src/Storages/MergeTree/CloudMergeTreeBlockOutputStream.cpp [hunk 2]
```diff
@@ -237,7 +237,7 @@ void CloudMergeTreeBlockOutputStream::writeSuffixImpl()
     {
         /// auto testlog = std::make_shared<TestLog>(const_cast<Context &>(context));
         /// TEST_START(testlog);
-        /// tryPreloadChecksumsAndPrimaryIndex(storage, std::move(preload_parts), ManipulationType::Insert, context);
+        tryPreload(context, storage, preload_parts);
```
Comment (hustnn) [inline src/Storages/MergeTree/CloudMergeTreeBlockOutputStream.cpp:240]: Do we need to support finish the preload before parts become visible?

File: src/CloudServices/commitCnchParts.cpp [hunk 3]
```diff
@@ -483,4 +484,37 @@ void CnchDataWriter::publishStagedParts(
     commitDumpedParts(items);
 }
 
+void tryPreload(
```
Comment (hustnn) [inline src/CloudServices/commitCnchParts.cpp:487]: How does mergemutate task preload the part？

File: src/Storages/MergeTree/CloudMergeTreeBlockOutputStream.cpp [hunk 4]
```diff
@@ -233,13 +233,7 @@ void CloudMergeTreeBlockOutputStream::writeSuffixImpl()
...
     }
 
-    if (!preload_parts.empty())
-    {
-        /// auto testlog = std::make_shared<TestLog>(const_cast<Context &>(context));
-        /// TEST_START(testlog);
-        /// tryPreloadChecksumsAndPrimaryIndex(storage, std::move(preload_parts), ManipulationType::Insert, context);
-        /// TEST_END(testlog, "Finish tryPreloadChecksumsAndPrimaryIndex in batch mode");
-    }
+    cnch_writer.preload(preload_parts);
```
Comment (hustnn) [inline src/Storages/MergeTree/CloudMergeTreeBlockOutputStream.cpp:236]: Is it duplicated? since it has been preloaded before commit.

Comment (Clark0) [reply src/Storages/MergeTree/CloudMergeTreeBlockOutputStream.cpp:236]: where do you refer to `the previous preload`? here it's actually preload after commit.

Comment (hustnn) [reply src/Storages/MergeTree/CloudMergeTreeBlockOutputStream.cpp]: I see the preload is also called such as commit part.

**Review by hustnn — APPROVED** [review]

**Pull request merged**